tests/e2e/csi_snapshot_negative.go

```diff
@@ -0,0 +1,887 @@
+
+	defer func() {
+		err := client.StorageV1().StorageClasses().Delete(ctx, storageclass.Name, *metav1.NewDeleteOptions(0))
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}()
+
+	ginkgo.By("Expect claim to provision volume successfully")
+	pvclaims = append(pvclaims, pvclaim)
+	persistentvolumes, err := fpv.WaitForPVClaimBoundPhase(client, pvclaims, framework.ClaimProvisionTimeout)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	volHandle := persistentvolumes[0].Spec.CSI.VolumeHandle
+	gomega.Expect(volHandle).NotTo(gomega.BeEmpty())
+
+	defer func() {
+		err := fpv.DeletePersistentVolumeClaim(client, pvclaim.Name, namespace)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		err = e2eVSphere.waitForCNSVolumeToBeDeleted(volHandle)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}()
+
+	// Verify using CNS Query API if VolumeID retrieved from PV is present.
+	ginkgo.By(fmt.Sprintf("Invoking QueryCNSVolumeWithResult with VolumeID: %s", volHandle))
+	queryResult, err := e2eVSphere.queryCNSVolumeWithResult(volHandle)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	gomega.Expect(queryResult.Volumes).ShouldNot(gomega.BeEmpty())
+	gomega.Expect(queryResult.Volumes[0].VolumeId.Id).To(gomega.Equal(volHandle))
+
+	ginkgo.By("Create volume snapshot class")
+	volumeSnapshotClass, err := snapc.SnapshotV1().VolumeSnapshotClasses().Create(ctx,
+		getVolumeSnapshotClassSpec(snapV1.DeletionPolicy("Delete"), nil), metav1.CreateOptions{})
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	framework.Logf("Volume snapshot class name is : %s", volumeSnapshotClass.Name)
+
+	defer func() {
+		err := snapc.SnapshotV1().VolumeSnapshotClasses().Delete(ctx,
+			volumeSnapshotClass.Name, metav1.DeleteOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	}()
+
+	ginkgo.By("Create a volume snapshot")
+	snapshot, err := snapc.SnapshotV1().VolumeSnapshots(namespace).Create(ctx,
+		getVolumeSnapshotSpec(namespace, volumeSnapshotClass.Name, pvclaim.Name), metav1.CreateOptions{})
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	framework.Logf("Volume snapshot name is : %s", snapshot.Name)
+	snapshotCreated = true
+
+	defer func() {
+		if snapshotCreated {
+			framework.Logf("Deleting volume snapshot")
+			err := snapc.SnapshotV1().VolumeSnapshots(namespace).Delete(ctx,
+				snapshot.Name, metav1.DeleteOptions{})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+			framework.Logf("Wait till the volume snapshot is deleted")
+			err = waitForVolumeSnapshotContentToBeDeleted(*snapc, ctx, snapshot.ObjectMeta.Name)
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			snapshotContentCreated = false
+		}
+
+		if snapshotContentCreated {
+			framework.Logf("Deleting volume snapshot content")
+			err := snapc.SnapshotV1().VolumeSnapshotContents().Delete(ctx,
+				*snapshot.Status.BoundVolumeSnapshotContentName, metav1.DeleteOptions{})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		}
+	}()
+
+	if serviceName == "CSI" {
```

Comment:
Can we put lines 329-432 in a different method as it is used several times in another function also.

Reply:
fixed it.